### PR TITLE
Choosing noise_kind and partition selection strategy in parameter tuning

### PR DESCRIPTION
**Pull request by dvadym**
This PR introduces `DPStrategySelector` class which finds `noise_kind` and partition selection strategy based on `eps`, `delta`, `max_partitions_contributed` etc. It does that by optimizing utility (e.g. `noise_kind` is chosen by minimum std deviation). 

Also this class is integrated into parameter tuning.

### CODE REVIEW

File: analysis/dp_strategy_selector.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+import pipeline_dp
+from pipeline_dp import aggregate_params
+from pipeline_dp import dp_computations
+from pipeline_dp import input_validators
+
+
+@dataclass
+class DPStrategy:
+    """Represents the chosen DP methods."""
```
Comment (miracvbasaran) [inline analysis/dp_strategy_selector.py:27]: nit: 'strategy' instead of 'methods'.

File: analysis/dp_strategy_selector.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    noise_kind: Optional[pipeline_dp.NoiseKind]
+    partition_selection_strategy: Optional[
+        pipeline_dp.PartitionSelectionStrategy]
+    post_aggregation_thresholding: bool
+
+
+class DPStrategySelector:
+    """Chooses DPStrategy based on DP budget, computation sensitivites etc
```
Comment (miracvbasaran) [inline analysis/dp_strategy_selector.py:35]: nit: ", etc."

File: analysis/dp_strategy_selector.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+    def __init__(self, epsilon: float, delta: float,
+                 metric: Optional[pipeline_dp.Metric],
+                 is_public_partitions: bool):
+        input_validators.validate_epsilon_delta(epsilon, delta,
+                                                "DPStrategySelector")
+        if delta == 0 and is_public_partitions:
+            raise ValueError(
+                "DPStrategySelector: when public_partitions are used, delta must be positive"
+            )
```
Comment (miracvbasaran) [inline analysis/dp_strategy_selector.py:51]: Why? Maybe they'll use Laplace & public partitions => delta = 0?

File: analysis/dp_strategy_selector.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        noise_kind = self.select_noise_kind(self._epsilon, self._delta,
+                                            sensitivities)
+        return DPStrategy(noise_kind=noise_kind,
+                          partition_selection_strategy=None,
+                          post_aggregation_thresholding=False)
+
+    def _get_dp_strategy_with_post_aggregation_threshold(
+            self, l0_sensitivity: int) -> DPStrategy:
+        assert self._metric == pipeline_dp.Metrics.PRIVACY_ID_COUNT
+        # Half delta goes to the noise, half for partition selection.
```
Comment (miracvbasaran) [inline analysis/dp_strategy_selector.py:97]: nit: 'Half of delta... , the other half for'

File: analysis/parameter_tuning.py [hunk 5]
```diff
@@ -179,6 +184,41 @@ def _find_candidate_parameters(
...
+        blueprint_params: pipeline_dp.AggregateParams,
+        noise_kind: Optional[pipeline_dp.NoiseKind],
+        strategy_selector: dp_strategy_selector.DPStrategySelector) -> None:
+    params = [
+        configuration.get_aggregate_params(blueprint_params, i)
+        for i in range(configuration.size)
+    ]
+    # if noise_kind is not None and strategy_selector.is_public_partitions:
+    #     # Noise kind and partition selection are already fixed.
+    #     return
```
Comment (miracvbasaran) [inline analysis/parameter_tuning.py:198]: Delete

File: analysis/parameter_tuning.py [hunk 6]
```diff
@@ -310,20 +352,39 @@ def tune(col,
         public_partitions: A collection of partition keys that will be present
           in the result. If not provided, tuning will be performed assuming
           private partition selection is used.
+        strategy_selector_factory: todo
```
Comment (miracvbasaran) [inline analysis/parameter_tuning.py:355]: Please add explanation

**Review by miracvbasaran — APPROVED** [review]

File: analysis/dp_strategy_selector.py (hunk 3 above)
Comment (dvadym) [reply analysis/dp_strategy_selector.py]: thanks, it's vice versa :)

**Review by dvadym — COMMENTED** [review]
Thanks for review!

**Pull request merged**